<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Артур, спасибо за проверку!!! И за допматериалы - всё унесла к себе, буду изучать. Я не справилась с bert - локально сейчас считает третий час и результата нет. Воспользуюсь твоим щедрым предложением идти дальше с текущей версией проекта - это проект жесткого дедлайна и без него не открывается теория. Спасибо!
</div>

### <font color='orange'>Общее впечатление (ревью 2)</font>
* После исправлений проект улучшился и теперь он может быть зачтен.
* Спасибо, что Обратила внимание на советы!
* Только Артем, а не Артур :)
* Удачи в дальнейшем обучении и следующих работах!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li><li><span><a href="#Общее-впечатление-(ревью-2)" data-toc-modified-id="Общее-впечатление-(ревью-2)-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span><font color="orange">Общее впечатление (ревью 2)</font></a></span></li></ul></li><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Сентимент-анализ-с-TF-IDF" data-toc-modified-id="Сентимент-анализ-с-TF-IDF-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Сентимент-анализ с TF-IDF</a></span></li><li><span><a href="#F1" data-toc-modified-id="F1-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>F1</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Линейная-регрессия" data-toc-modified-id="Линейная-регрессия-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Линейная регрессия</a></span></li><li><span><a href="#Градиентный-бустинг-с-CatBoost" data-toc-modified-id="Градиентный-бустинг-с-CatBoost-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Градиентный бустинг с CatBoost</a></span></li></ul></li><li><span><a href="#Тестирование" data-toc-modified-id="Тестирование-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Тестирование</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Здравствуй, дорогой ревьювер! Этот проект стал для меня настоящим "сыном ошибок трудных". Вроде бы он не очень сложен и ничто не предвещало беды - но я сделала несколько досадных ошибок и на их поиски потратила 4 дня. А если учесть, что считает очень подолгу, ядро регулярно умирает, и на подбор ушло еще четыре, и я выбилась из графика окончательно - это так я объясняю, почему тут нет bert. Но я постараюсь её доделать и включить.   </div></font>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

Импортируем библиотеки:

In [63]:
import numpy as np
import pandas as pd
import torch
import transformers
import re
import warnings

from tqdm import notebook
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import f1_score
from sklearn.metrics import make_scorer 
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression
from catboost import Pool, CatBoostClassifier, cv

import nltk
from nltk.corpus import stopwords
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')
nltk.download('punkt')
nltk.download('omw-1.4')
from nltk.stem import WordNetLemmatizer 
from nltk.corpus import wordnet
from tqdm import notebook


[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/jovyan/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


Скрытие предупреждений

In [64]:
pd.options.mode.chained_assignment = None
warnings.filterwarnings('ignore')

Загрузка данных

In [66]:
data = pd.read_csv('/datasets/toxic_comments.csv', index_col = [0])

In [67]:
data.to_csv('toxic_comments.csv')

Константы

In [4]:
RANDOM_STATE = 12345

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 159292 entries, 0 to 159450
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 3.6+ MB


Таблица комментариев представлена двумя полностью заполненными столбцами. Наименования столбцов латиницей в нижнем регистре, имеют смысловую связь с содержанием. Столбец text содержит текст комментария, а toxic — целевой признак.  Был столбец, дублирующий индексы, переоткрыли правильно. Пропусков нет. Посмотрим на значения столбца toxic:

In [6]:
data['toxic'].value_counts()

0    143106
1     16186
Name: toxic, dtype: int64

Интересно, а о чём пишут?

In [7]:
data[data['toxic'] == 1].head()

text  toxic
6        COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK      1
12  Hey... what is it..\n@ | talk .\nWhat is it......      1
16  Bye! \n\nDon't look, come or think of comming ...      1
42  You are gay or antisemmitian? \n\nArchangel WH...      1
43           FUCK YOUR FILTHY MOTHER IN THE ASS, DRY!      1

Ой вэй... точно не про продукцию. А светлая сторона о чём?

In [8]:
data[data['toxic'] == 0].head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

Как-то даже не верится, что там что-то  о продукции - людям явно интереснее люди. Так, а какова доля негатива?

In [9]:
data[data['toxic'] == 1].count() / data.shape[0]*100

text     10.161213
toxic    10.161213
dtype: float64

Десять процентов токсичных коментариев. Довольно умеренная жестокость. Сделаем её контролируемой.

**Вывод** Таблица комментариев представлена двумя полностью заполненными столбцами. Наименования столбцов латиницей в нижнем регистре, имеют смысловую связь с содержанием. Столбец text содержит текст комментария, а toxic — целевой признак. Пропусков нет. Данные готовы к обработке.

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корреткно. Радует, что баланс классов был изучен.
</div>

## Обучение

### Сентимент-анализ с TF-IDF

Нам нужно определить тональность текста, и в нашем случае негативные комментарии уже помечены.


In [10]:
#corpus = data['text'].values.astype('U')
#corpus

Произведем лемматизацию — приведение слова к начальной форме — лемме, учитывая, что у нас английский язык.

Чистим от лишних символов модулем re (сокр. от regular expressions):



Создадим корпус: преобразуем столбец text в список текстов.

In [11]:
def clear_text(row):
    cleared_text = re.sub(r'[^a-zA-Z]', ' ', row['text'])
    cleared_text = " ".join(cleared_text.split())
    return cleared_text

Удалаем лишние пробелы, применяя комбинацию функций join() и split(). Методом split() преобразуем этот текст в список. 

In [12]:
#data = data.sample(400)
data['text']= data.apply(clear_text, axis = 1)
data.head()

text  toxic
0  Explanation Why the edits made under my userna...      0
1  D aww He matches this background colour I m se...      0
2  Hey man I m really not trying to edit war It s...      0
3  More I can t make any real suggestions on impr...      0
4  You sir are my hero Any chance you remember wh...      0

Чистейший текст. Может быть верхний апостроф нужно было оставить? Он смысловой...
Переходим к лемматизации:

In [13]:
lemmatizer = WordNetLemmatizer()
def get_wordnet_pos(word):
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)
    
    lemm_list = [lemmatizer.lemmatize(w, get_wordnet_pos(w)) for w in nltk.word_tokenize(sentence)]
    lemm_text = "".join(lemm_list)
        
    return lemm_text

def lemmatize(row):
    sentence = row['text']
    lemmas = " ".join([lemmatizer.lemmatize(w, get_wordnet_pos(w)) for w in nltk.word_tokenize(sentence)])
    return lemmas

Применим функцию лемматизации к нашему столбцу текстов:

In [14]:
data['text']= data.apply(lemmatize, axis = 1)
data.head()

text  toxic
0  Explanation Why the edits make under my userna...      0
1  D aww He match this background colour I m seem...      0
2  Hey man I m really not try to edit war It s ju...      0
3  More I can t make any real suggestion on impro...      0
4  You sir be my hero Any chance you remember wha...      0

После этой операции в тексте можно обнаружить лишние пробелы, для анализа они — помеха. Пробелы устраняются комбинацией функций join(). Ещё раз очистим.

In [15]:
data['text']= data.apply(clear_text, axis = 1)
data.head()

text  toxic
0  Explanation Why the edits make under my userna...      0
1  D aww He match this background colour I m seem...      0
2  Hey man I m really not try to edit war It s ju...      0
3  More I can t make any real suggestion on impro...      0
4  You sir be my hero Any chance you remember wha...      0

TF IDF и линейная регрессия

In [16]:
#corpus = data['lemm_text'].values.astype('U')


Разбиваем датасет на признаки и целевой признак

In [17]:
features = data.drop(['toxic'], axis=1)
target = data['toxic']
features

text
0       Explanation Why the edits make under my userna...
1       D aww He match this background colour I m seem...
2       Hey man I m really not try to edit war It s ju...
3       More I can t make any real suggestion on impro...
4       You sir be my hero Any chance you remember wha...
...                                                   ...
159446  And for the second time of ask when your view ...
159447  You should be ashamed of yourself That be a ho...
159448  Spitzer Umm there no actual article for prosti...
159449  And it look like it be actually you who put on...
159450  And I really don t think you understand I come...

[159292 rows x 1 columns]

In [18]:
corpus_train, corpus_test, target_train, target_test = train_test_split( features, target, test_size=0.2, random_state=RANDOM_STATE)
corpus_train

text
97494   Bushranger you re a GRASS with no sense of hum...
4383    Need administrative help I have be block iniqu...
103777  I d also like to point out that he have use a ...
38619        You cant block me you fuck retard BRB nigger
128443  I believe that the frequency of the wave need ...
...                                                   ...
110090  Hahaha I dont live in a lie like you and dont ...
85493                                         March March
133387  Agreed We really should try to stick to the su...
130469  Umm killer Do you not like that he copy your w...
77361      Bradford City I be remove unreferanced content

[127433 rows x 1 columns]

In [19]:
stopwords = set(stopwords.words('english'))

count_tf_idf = TfidfVectorizer(stop_words=stopwords)

tf_idf_train = count_tf_idf.fit_transform(corpus_train['text'])
print(tf_idf_train.shape)
tf_idf_train


(127433, 139430)


<127433x139430 sparse matrix of type '<class 'numpy.float64'>'
	with 3360801 stored elements in Compressed Sparse Row format>

In [20]:
tf_idf_test = count_tf_idf.transform(corpus_test['text'])

print(tf_idf_test.shape)

(31859, 139430)


In [21]:
model = LogisticRegression(random_state=12345, solver="lbfgs")
model.fit(tf_idf_train, target_train)
pred_train = model.predict(tf_idf_train)


In [22]:
f1_l_train = f1_score(target_train, pred_train)
f1_l_train

0.7593974507531865

In [23]:
pred_test = model.predict(tf_idf_test)
pred_test = pd.DataFrame(data=pred_test, columns=['toxic'])



In [24]:
f1_l_test = f1_score(target_test, pred_test)
f1_l_test

0.7379039790771529

### F1 

In [25]:
from sklearn.metrics import make_scorer

def f1_scor(target, predictions):
    f1 = f1_score(target, predictions)
    return f1

f1_scorer = make_scorer(f1_scor, greater_is_better=True)


### Дерево решений

Первая модель  - дерево решений. Подберём глубину кросс-валидацией.

In [32]:
best_model = None
best_cross_tr = 0
best_depth = 0
for depth in range(151, 262, 30): # Проверялся диапазон 1 351.
    model_tr =  DecisionTreeClassifier(random_state=RANDOM_STATE, max_depth=depth) 
    model_tr.fit(tf_idf_train, target_train) 
    predictions_train = model_tr.predict(tf_idf_train) 
    cross = cross_val_score(model_tr, tf_idf_train, target_train, cv=5, scoring = f1_scorer)
    if cross.mean()> best_cross_tr:
        best_model = model_tr
        best_cross_tr = cross.mean()
        best_depth = depth
        print(best_cross_tr,  depth)

print("f1 наилучшей модели на обучающей выборке:", best_cross_tr, "Глубина дерева:", best_depth)

0.7185606973228214 151
0.7200896722973104 181
f1 наилучшей модели на обучающей выборке: 0.7200896722973104 Глубина дерева: 181


Перебор в диапазоне 1 - 61 показал:- 0.7027939952359202 61,
0.7045398159782007 66,
0.70484012064318 71,
0.7072376798303114 76, 
0.7082912202428133 81, 
0.7094674015913008 86, 
0.7123389600870491 91,
0.7166630772025926 111
0.7184806097501119 116
0.7189559351464924 121
0.7118565811292771 351
0.7185606973228214 151
0.7200896722973104 181

Лучший результат f1 на обучающей выборке: 0.7200896722973104 Глубина дерева: 181.

In [33]:
%%time
model_tr =  DecisionTreeClassifier(random_state=RANDOM_STATE, max_depth=181) # инициализируйте модель DecisionTreeClassifier с параметром random_state=12345 и max_depth=depth
model_tr.fit(tf_idf_train, target_train) 
predictions_train = model_tr.predict(tf_idf_train) 
predictions_train

CPU times: user 1min 42s, sys: 293 ms, total: 1min 42s
Wall time: 1min 42s


array([1, 0, 0, ..., 0, 0, 0])

Ограничения по времени у нас нет (упоминаю об этом, поскольку считает долго), получена приемлемая величина метрики f1 - 0,72 на кросс-валидации.

### Случайный лес

Следующая модель  - случайный лес. Снова кросс-валидация.

In [36]:
best_model = None
best_est = 0 
best_depth = 0 
best_cross_rfr = 0

for est in range(36, 38, 3): #был диапазон (1, 45), сокращён до близкого к лучшим значениям
    for depth in range(351,381, 10): #был диапазон (1, 451), сокращён до близкого к лучшим значениям

        model_f = RandomForestClassifier(n_estimators=est, 
                                        max_depth=depth, 
                                        random_state=RANDOM_STATE,  
                                        ) 
        model_f.fit(tf_idf_train, target_train)
        prediction_f = model_f.predict(tf_idf_train)
        cross_rfr = cross_val_score(model_f, tf_idf_train, target_train, cv=5, scoring = f1_scorer) 
        if  cross_rfr.mean()> best_cross_rfr:
            best_est = est
            best_depth = depth
            best_cross_rfr = cross_rfr.mean()
            best_model = model_f
            print(cross_rfr.mean(), est, depth)
print(best_cross_rfr, 'Лучшие гиперпараметры:', best_est,  best_depth)


0.630413936165651 36 351
0.6347776503985033 36 361
0.6347776503985033 Лучшие гиперпараметры: 36 361


0.5274583282966828 30 210
0.5365899775692459 30 220
0.5499613482814882 30 230
0.630413936165651 36 351
0.6347776503985033 36 361

Лучший средний результат f1 на кросс- валидации - 0.6347776503985033 Лучшие гиперпараметры модели случайный лес: n_estimators - 36,  max_depth - 361.

In [37]:
model_rfr = RandomForestClassifier(n_estimators=37,  max_depth=361,  random_state=RANDOM_STATE )
model_rfr.fit(tf_idf_train, target_train)

RandomForestClassifier(max_depth=361, n_estimators=37, random_state=12345)

In [38]:
predictions_rfr = model_rfr.predict(tf_idf_train)

predictions_rfr

array([1, 0, 0, ..., 0, 0, 0])

### Линейная регрессия

Двигаемся дальше, следующая по плану - линейная регрессия.

In [42]:
model_l = LogisticRegression(max_iter=500, 
                           C=0.1, 
                           solver='lbfgs', 
                           n_jobs=-1, 
                           random_state=RANDOM_STATE, 
                           class_weight='balanced')

In [43]:
model_l.fit(tf_idf_train, target_train)

LogisticRegression(C=0.1, class_weight='balanced', max_iter=500, n_jobs=-1,
                   random_state=12345)

In [44]:
predictions_train_l = model_l.predict(tf_idf_train) # получим предсказания модели на тренировочной выборке
cross_l = cross_val_score(model_l, tf_idf_train, target_train, cv=5, scoring = f1_scorer) 

print("F1, кросс-валидация:", cross_l.mean())


F1, кросс-валидация: 0.7003074790682955


In [45]:
predictions_train_l.min()

0

In [46]:
predictions_train_l.max()

1

In [ ]:
#predictions_train_l = pd.DataFrame(predictions_train_l, columns=['toxic']) 
#predictions_train_l

### Градиентный бустинг с CatBoost

In [53]:
model_cat = CatBoostClassifier(random_state=RANDOM_STATE,  loss_function='Logloss', iterations=300)
model_cat.fit(tf_idf_train, target_train,verbose=10, plot=True)

Custom logger is already specified. Specify more than one logger at same time is not thread safe.

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

Learning rate set to 0.246246
0:	learn: 0.4783848	total: 3.19s	remaining: 15m 53s
10:	learn: 0.2126400	total: 25.5s	remaining: 11m 8s
20:	learn: 0.1887209	total: 46.5s	remaining: 10m 17s
30:	learn: 0.1766898	total: 1m 7s	remaining: 9m 49s
40:	learn: 0.1680734	total: 1m 29s	remaining: 9m 22s
50:	learn: 0.1613705	total: 1m 49s	remaining: 8m 55s
60:	learn: 0.1550557	total: 2m 10s	remaining: 8m 32s
70:	learn: 0.1509122	total: 2m 31s	remaining: 8m 8s
80:	learn: 0.1471269	total: 2m 52s	remaining: 7m 46s
90:	learn: 0.1433277	total: 3m 14s	remaining: 7m 27s
100:	learn: 0.1404027	total: 3m 34s	remaining: 7m 3s
110:	learn: 0.1375419	total: 3m 55s	remaining: 6m 41s
120:	learn: 0.1350952	total: 4m 16s	remaining: 6m 20s
130:	learn: 0.1332454	total: 4m 37s	remaining: 5m 57s
140:	learn: 0.1311162	total: 4m 58s	remaining: 5m 36s
150:	learn: 0.1289473	total: 5m 19s	remaining: 5m 15s
160:	learn: 0.1270528	total: 5m 40s	remaining: 4m 54s
170:	learn: 0.1254366	total: 6m	remaining: 4m 32s
180:	learn: 0.123

In [54]:
predictions_cat_train = model_cat.predict(tf_idf_train)
cross_cat = cross_val_score(model_cat, tf_idf_train, target_train, cv=5, scoring = f1_scorer) 

print("F1, кросс-валидация c CatBoost на обучении:", cross_cat.mean())


Learning rate set to 0.223866
0:	learn: 0.4869524	total: 2.36s	remaining: 11m 45s
1:	learn: 0.3750135	total: 4.2s	remaining: 10m 26s
2:	learn: 0.3135108	total: 6.01s	remaining: 9m 55s
3:	learn: 0.2778402	total: 7.81s	remaining: 9m 37s
4:	learn: 0.2552028	total: 9.6s	remaining: 9m 26s
5:	learn: 0.2420909	total: 11.4s	remaining: 9m 19s
6:	learn: 0.2333580	total: 13.2s	remaining: 9m 11s
7:	learn: 0.2271543	total: 15s	remaining: 9m 6s
8:	learn: 0.2211772	total: 16.8s	remaining: 9m 1s
9:	learn: 0.2172942	total: 18.7s	remaining: 9m 1s
10:	learn: 0.2136719	total: 20.4s	remaining: 8m 57s
11:	learn: 0.2105311	total: 22.2s	remaining: 8m 52s
12:	learn: 0.2072362	total: 23.9s	remaining: 8m 47s
13:	learn: 0.2041333	total: 25.8s	remaining: 8m 46s
14:	learn: 0.2017173	total: 27.6s	remaining: 8m 43s
15:	learn: 0.1996930	total: 29.3s	remaining: 8m 40s
16:	learn: 0.1981417	total: 31.2s	remaining: 8m 38s
17:	learn: 0.1956622	total: 33s	remaining: 8m 36s
18:	learn: 0.1939199	total: 34.8s	remaining: 8m 34s

F1, среднее значение кросс-валидация c CatBoost на обучении: 0.7342181781066941

In [40]:
predictions_rfr_test = model_rfr.predict(tf_idf_test)
f1_rfr_test = f1_score(target_test, predictions_rfr_test)
f1_rfr_test

0.6281678437889489

In [55]:
predictions_cat_train

array([1, 0, 0, ..., 0, 0, 0])

In [56]:
predictions_cat_train.min()

0

In [57]:
predictions_cat_train.max()

1

Переходим к выбору лучшей модели и тестированию.

## Тестирование

Сравним полученные результаты:

In [61]:
data = {'F1 на обучении ':[ best_cross_tr, best_cross_rfr, cross_l.mean(), cross_cat.mean()]} 
# Creates pandas DataFrame. 
plt = pd.DataFrame(data, index =[ 'Дерево решений', 'Случайный лес', 'Линейная регрессия',
                                  'Градиентный бустинг с CatBoost']) 

display(plt)

F1 на обучении 
Дерево решений                         0.720090
Случайный лес                          0.634778
Линейная регрессия                     0.700307
Градиентный бустинг с CatBoost         0.734218

При сравнении получается, что все модели показали относительно высокие результаты, но есть бесспорный лидер - Это Градиентный бустинг с CatBoost и есго средним значением метрики F1 на кросс-валидации 0.7342. Его мы и оставим для теста. 

Прекдсказываем с помощью CatBoost на тестовых данных:

In [62]:
predictions_cat_test = model_cat.predict(tf_idf_test)

f1_l_test = f1_score(target_test, predictions_cat_test)
f1_l_test

0.7522139887945057

## Выводы

**Вывод** Лучшей моделью по итогам подбора параметров при помощи кросс-валидации стала модель Градиентный бустинг с CatBoost. Наша модель удовлетворяет условию задания -  модель со значением метрики качества F1 не меньше 0.75: значение метрики F1 на тестовой выборке - 0.7522. 

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны